File: iosApp/iosApp/Modules/Alert/AlertWireframeFactory.swift
// Created by web3dgn on 02/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

protocol AlertWireframeFactory {

    func make(_ parent: UIViewController?, context: AlertWireframeContext) -> AlertWireframe
}

final class DefaultAlertWireframeFactory {}

extension DefaultAlertWireframeFactory: AlertWireframeFactory {
    
    func make(_ parent: UIViewController?, context: AlertWireframeContext) -> AlertWireframe {
        DefaultAlertWireframe(parent: parent, context: context)
    }
}

// MARK: - Assembler

final class AlertWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { _ -> AlertWireframeFactory in
            DefaultAlertWireframeFactory()
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/alert/AlertWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.alert

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.alert.AlertWireframe
import com.sonsofcrypto.web3walletcore.modules.alert.AlertWireframeContext
import smartadapter.internal.extension.name

interface AlertWireframeFactory {
    fun make(parent: Fragment?, context: AlertWireframeContext): AlertWireframe
}

class DefaultAlertWireframeFactory: AlertWireframeFactory {

    override fun make(parent: Fragment?, context: AlertWireframeContext): AlertWireframe {
        return DefaultAlertWireframe(
            parent?.let { WeakRef(it) },
            context
        )
    }
}

class AlertWireframeAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(AlertWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultAlertWireframeFactory()
        }
    }
}